### ISSUE TITLE
Heroku is no more providing free plan: we need a new CD platform

### ISSUE THREAD

**Issue by dduportal** [cicd-lectures/slides#89]
As per https://techcrunch.com/2022/08/25/heroku-announces-plans-to-eliminate-free-plans-blaming-fraud-and-abuse/, we can't use Heroku for free anymore.

We need a new solution: GCP Cloudrun maybe? CleverCloud?

**Comment by jlevesy** [comment]
Okay I've run an initial test, here are my notes:

- The menu-server, as it is, does not need to  be adapted in order to run on cloud-run (all it needs respond to ${PORT})
- How it roughly works:
  - We need to build and push an image on the google artifact registry (`docker build eu.gcr.io/voi-sdbx-jly-test-cloudrun/jlevesy/menu-server:test-cloudrun && docker push`)
  - From there we declare a cloud run deployment referencing the newly created image with the following:
  
```
 gcloud run deploy \
     menu-server-jlevesy \
     --image=eu.gcr.io/voi-sdbx-jly-test-cloudrun/jlevesy/menu-server:test-cloudrun \ 
     --region=europe-west6  \
     --min-instances=1 \ 
     --max-instances=1 \
     --project=voi-sdbx-jly-test-cloudrun \
    --allow-unauthenticated
```

  - And there you have it. A fully functional web app and an URL returned in the output.
  - And you can get rid of it using `gcloud run services delete menu-server-jlevesy --project voi-sdbx-jly-test-cloudrun --region=europe-west6`

Cool, great, I think that will do nicely, but that was the easy part now I need to:

- Find a way to provision the GCP project so that students can log-in with gcloud from their codespace / gitpod and run those commands. With the minimal perms allowed of course, so that it is mistake proof (and doesn't cost me my job 😅). The first thing I'm going to look for is a way for them to log in and authenticate, preferably without a google account, even if I'm pretty sure that's going to be impossible :D.  So yes **they will probably need a google account**.
- Also find a way to make this work with GitHub actions, but in [that case workload identity federation with github OIDC](https://cloud.google.com/iam/docs/workload-identity-federation) will do the trick, I'll only need to build an allow list of their repositories.
- Google cloud provide a set of [actions](https://github.com/google-github-actions/deploy-cloudrun#usage) that are going to help us. 

TL;DR; next step: figure out the IAM setup and authentication for students to allow them to push images and deploy on cloudrun.

**Comment by jlevesy** [comment]
So here's what I have:

- I've created a Terraform project with all the necessary setup to have students manipulate cloud run from their workflow, or from their console (I have yet to test this though 🙃) See this [PR](https://github.com/cicd-lectures/gcp-cd-infra/pull/1)
- I also (after a fight) have updated the example github worklflow to have it release a new version on cloud run, my notes about that:
  - The [job I have right now](https://github.com/cicd-lectures/menu-server/blob/2023/.github/workflows/ci.yml#L64) is much much more complicated than the one with Heroku. 
  - We either need to give the solution piece by piece, or find another way. Because as it is, they will never find the solution by themselves. This is mostly due to the fact that we're both using artifact registry and OIDC auth :*( 
  - I didn't foresee that adding OIDC will make it that difficult 😞. I'll try to make it as bearable in the slides as possible, we don't need to to explain it too much. 
  - We [can't allow unauthorized calls to our cloud run instances from github actions](https://github.com/google-github-actions/deploy-cloudrun#allowing-unauthenticated-requests)  but I'm testing a hack to bypass that. Otherwise we would have to fix it by hand after they have created their instance. 

My next steps are:
- Test also test this  setup from gitpod to see if it works.
  - I need  https://github.com/cicd-lectures/gitpod/pull/5 to land before. 
- Update the slides.

I would be happy to get your opinion @dduportal! Happy to chat about it!

**Comment by jlevesy** [comment]
Setup instructions for gcloud:

- `gcloud auth login`
- `gcloud config set project <project-id>`
- `gcloud auth configure-docker europe-west9-docker.pkg.dev`

But overall it's working from git pod as well, we only have to add the students google accounts to our google group to grant them access :)

**Comment by jlevesy** [comment]
And I have written an action that simplifies a bit the final release job. [here's an usage example](https://github.com/cicd-lectures/menu-server/blob/2023/.github/workflows/ci.yml#L79) let me know what you think about that.

I think it can be simplified even more.

**Comment by dduportal** [comment]
As we tested #93, everything works flawlessly with GCP. We only need to refine the chapter steps.
